pageserver/src/page_cache.rs

```diff
@@ -308,7 +307,9 @@ impl Drop for PageWriteGuard<'_> {
         assert!(self.inner.key.is_some());
         if !self.valid {
             let self_key = self.inner.key.as_ref().unwrap();
-            PAGE_CACHE.get().unwrap().remove_mapping(self_key);
+            tokio::task::block_in_place(|| {
```

Comment:
block_in_place is super expensive, we should avoid this.

Alternatives:
- We could carry an mpsc UnboundedChannel Sender in PageWriteGuard, to send notifications for the page cache to consume, to release such slots asynchronously
- We could carry a dirty bit in Slot or SlotInner, and the page cache can clean it up next time someone accesses the slot, or if the Clock cursor sees it.

Reply:
There may be a higher level answer than this though, like making page contents CoW so that a reader can hold a ref to some contents to read without having to hold a lock on a page cache slot

Reply:
I had a call yesterday with @koivunej about this question. There is two usages of `block_in_place` in this PR, both due to accesses in `Drop` impls that access the page cache. The first usage is in the EphemeralFile destructor, to free all pages associated with the to-be-deleted ephemeral file. The second usage is in the destructor of the page write guard, and only frees that single page. In the call, we found that `EphemeralFile` performs a complex search over the page cache for the pages to free, so we probably can't use dirty bits here. For the `PageWriteGuard` however, we discovered something  we could use as "dirty bit": the `Slot::usage_count` value, which is used in the `find_victim` function for determining whether we want to remove a specific page.

Since the call I figured out that the problem with that value is that the `PageWriteGuard` does not have a reference to `Slot` but `SlotInner`, and `usage_count` is stored in former :/.

I've been wondering about adding a `&'static Slot` field in `SlotInner`, but initialization proved to be a problem: first, one needs to make it an `Option` to prevent a circular dependency: one can first initialize all the `Slot`s with `SlotInner`s that have `None` for the pointer, and then set the pointer to `Some` in a later step. The only issue is just where to do that, as `PageCache::new` is called in settings that don't support async even with this PR, and the `SlotInner` is behind a lock, locking which would become an async operation in the future.

Adding a `to_clear` field to `SlotInner` also runs into problems, as `find_victim` doesn't actually lock the `SlotInner` values except unless it found an actual victim. Locking just to read `to_clear` is a big no-no.

Adding the `Slot` pointer to the write guard is also non-optimal as it increases the size of the guard.

One could just do the thing that people in Rust land do when pointers become too much of a trouble: use indices instead :).

Or maybe one could just always use queues. Ideally though, I want as little overhead to exist for `PageWriteGuard` dropping. Still thinking...

Reply:
I have ended up using `slot_idx` for that purpose, it works.

I also looked into the channel implementations:

* `std::sync::mpsc` channels are not `Sync` for the `Receiver` type
* `tokio::sync::mpsc` channels are `Sync` but the receiver takes a `&mut self` reference. all the page cache functions deal with a `&self` page cache...
* `crossbeam_channel` channels have `Sync` both for `Sender` and `Receiver` types, and take `&self` types.

So I suppose we'll have to go with `crossbeam_channel` channels.

Reply:
This issue is resolved now by not using tokio locks for the maps.